woocommerce/woocommerce-ios
# Use NSPersistentStoreCoordinator Instead of FileManager to Migrate Stores

Closes #2667. This is a continuation of #3486. 

This is the last step in fully removing the usage of `FileManager` to migrate stores. As we've seen observed with the effect of https://github.com/woocommerce/woocommerce-ios/pull/3069, removing the `FileManager` is effective in reducing the crashes. This is the from the [Sentry event](https://sentry.io/share/issue/d1497b3ad58e4cba8249599fd9efc0eb/):

<img src="https://user-images.githubusercontent.com/198826/100031853-e20f9500-2db3-11eb-91e1-abaa55895076.png" width="300">

Hopefully, this change fully eliminates all of them. 🤞 

## Changes

### Destroying and Migrating

This primarily removes the `FileManager` operations like this: 

https://github.com/woocommerce/woocommerce-ios/blob/5cc71164342b0838c353fe9add9e5e5504bf3c8d/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L186-L189

And replaces them with `NSPersistentStoreCoordinator` functions like [`destroyPersistentStore`](https://developer.apple.com/documentation/coredata/nspersistentstorecoordinator/1468888-destroypersistentstore) and [`replacePersistentStore`](https://developer.apple.com/documentation/coredata/nspersistentstorecoordinator/1468917-replacepersistentstore). 

### Temporary Files

Previously, we use the `Documents` directory for temporary store files. 

https://github.com/woocommerce/woocommerce-ios/blob/5cc71164342b0838c353fe9add9e5e5504bf3c8d/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L93-L99

This changes that so that the temporary directory will be used instead:

https://github.com/woocommerce/woocommerce-ios/blob/b1f3fe11610f34913ab554491946cdaaa5a15726/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L171-L175

This helps in avoiding a lot of file operations. And we also don't need to fully clean up everything if the app crashes in between. 

### Fewer Store Replacements

Currently, every incremental migration will replace the store URL (`~/Documents/WooCommerce.sqlite`). 

https://github.com/woocommerce/woocommerce-ios/blob/5cc71164342b0838c353fe9add9e5e5504bf3c8d/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L188-L189

For example, when migrating from Model 39 to Model 41, the store URL will be replaced twice. 

This PR changes that so that all incremental migrations will have to be finished first before the store URL [gets replaced](https://github.com/woocommerce/woocommerce-ios/blob/b1f3fe11610f34913ab554491946cdaaa5a15726/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L94-L103). Using the same example above, migrating from Model 39 to Model 41 will only cause the store URL to be replaced **once**. If an error happens in between, there are fewer chances that the store URL will be corrupted. 

### Flow 

Hopefully, these flow diagrams will better explain what I mentioned above. 😅 

#### Before

<img src="https://user-images.githubusercontent.com/198826/105790982-7b892d80-5f42-11eb-8720-19a698679471.png" width="500">

#### After

<img src="https://user-images.githubusercontent.com/198826/105790944-6dd3a800-5f42-11eb-8f13-5be4a7d03ab4.png" width="700">


## Testing

Please do a regression test that migrations still happen.

1. Delete the app on the device. 
2. Using this branch, delete models 41 and 42. 
3. Set the current model version to 40. 
4. Run the app and log in. But probably don't open Orders and Products because we're missing some Core Data properties. 😅 
5. Stop the app. 
6. Restore the deleted model versions and reset the current model version to 42. 
7. Add a breakpoint on this line:

    https://github.com/woocommerce/woocommerce-ios/blob/11ed114d61b80b9332d97cf598e9a1a52e8cfcfe/Storage/Storage/CoreData/CoreDataIterativeMigrator.swift#L76
8. Add a breakpoint on this line:

    https://github.com/woocommerce/woocommerce-ios/blob/11ed114d61b80b9332d97cf598e9a1a52e8cfcfe/Storage/Storage/CoreData/CoreDataManager.swift#L79

9. Run the app again. Migrations should happen. Confirm that the breakpoint on Step 7 was hit twice. And the `migrationAttemptMessage` variable contain these values:
    - `⚠️ Attempting migration from Model 40 to Model 41`
    - `⚠️ Attempting migration from Model 41 to Model 42`
10. Confirm the breakpoint on Step 8 was never hit. 
11. Confirm that you can load the Orders and Products. 


## Author Checklist
